File: frontend/lib/api/analytics.ts
import { apiClient } from '@/lib/api/client';

export interface InstallmentDetail {
  description: string;
  monthly_payment: number;
  remaining_months: number;
}

export interface InstallmentAnnotation {
  description: string;
  category_name: string | null;
  monthly_payment: number;
  original_amount: number;
  remaining_amount: number;
  started_this_month: boolean;
}

export interface CategoryExpense {
  category_id: number | null;
  category_name: string;
  amount: number;
  is_regular: boolean;
  installment_details: InstallmentDetail[] | null;
}

export interface InstallmentAccountSummary {
  account_name: string;
  total_debt: number;
  monthly_payment: number | null;
  has_purchase_details: boolean;
}

export interface ExpenseAnalytics {
  total_expenses: number;
  regular_expenses: number;
  irregular_expenses: number;
  categories: CategoryExpense[];
  installment_annotations: InstallmentAnnotation[];
  new_installment_obligations: number;
}

export async function getExpenseAnalytics(year: number, month: number): Promise<ExpenseAnalytics> {
  return apiClient<ExpenseAnalytics>(`/analytics/expenses?year=${year}&month=${month}`);
}

File: app/schemas/analytics.py
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel


class InstallmentDetail(BaseModel):
    description: str
    monthly_payment: Decimal
    remaining_months: int


class InstallmentAnnotation(BaseModel):
    description: str
    category_name: str | None
    monthly_payment: Decimal
    original_amount: Decimal
    remaining_amount: Decimal
    started_this_month: bool


class CategoryExpense(BaseModel):
    category_id: int | None
    category_name: str
    amount: Decimal
    is_regular: bool
    installment_details: list[InstallmentDetail] | None = None


class InstallmentAccountSummary(BaseModel):
    account_name: str
    total_debt: Decimal
    monthly_payment: Decimal | None = None
    has_purchase_details: bool


class ExpenseAnalyticsResponse(BaseModel):
    total_expenses: Decimal
    regular_expenses: Decimal
    irregular_expenses: Decimal
    categories: list[CategoryExpense]
    installment_annotations: list[InstallmentAnnotation]
    new_installment_obligations: Decimal
    installment_accounts_summary: list[InstallmentAccountSummary] = []
